### ISSUE TITLE
Network scaling

### ISSUE THREAD

**Issue by DeepaMahm** [networkit/networkit#600]
Hello Everyone,

This is  a follow up to my previous post [here](https://stackoverflow.com/questions/64131201/methods-for-generating-complex-network/64150221#64150221)

I'm looking for network scaling algorithms in Networkit. The section on `The generation algorithm ReCoN` in [this ](https://link.springer.com/article/10.1007/s41109-017-0054-z) article presents a sequence of steps and in the article, it is mentioned that ` The ReCoN code is publicly available in the open-source network analysis package NetworKit`. Could someone help me in locating the ReCoN or the function that can scale user-defined graph networks?

**Comment by hmeyerhenke** [comment]
Look for the fit() classmethod of the LFRGenerator class in networkit/generators.pyx. ReCoN is only available via that Python code.

**Comment by DeepaMahm** [comment]
Hi @hmeyerhenke 

Thanks a lot.
I tried the below,

```
import networkit as nk
import random
import networkx as nx
import matplotlib.pyplot as plt
from networkit.generators import LFRGenerator

G = nx.gnm_random_graph(20, 30, seed=1)

for (u, v) in G.edges():
    G.edges[u, v]['weight1'] = random.randint(0, 10)
    G.edges[u, v]['weight2'] = random.randint(0, 10)

nx.draw(G, with_labels=True)
plt.show()
print(G.edges(data=True))

# Networkx graph to NetworKit graph
G_nk = nk.nxadapter.nx2nk(G, weightAttr='weight1')
# print(G_nk)
# print(G_nk.weight(u=0, v=12))

LFRGenerator.fit(G_nk, scale=2)

```
I'm not sure how to access the contents of what's returned by this class method `LFRGenerator.fit(G_nk, scale=2)`. 
I tried,

```
G_nk_scaled = LFRGenerator.fit(G_nk, scale=2)
print(G_nk_scaled.nodes()) 

```
to check if the number of nodes in the scaled graph is doubled. Unfortunately, that doesn't work.

**Comment by DeepaMahm** [comment]
@hmeyerhenke This is a kind reminder

**Comment by hmeyerhenke** [comment]
See the replies by @michitux and @cndolo.

**Comment by bielefem** [comment]
@DeepaMahm

tldr: look at the bottom

It looks like you assumed to that LFRGenerator.fit(G_nk, scale=2) returns a graph,
given the evidence of your snippet:
```python
G_nk_scaled = LFRGenerator.fit(G_nk, scale=2)
print(G_nk_scaled.nodes()) 
```

a quick look at the return type help us to see that it returns an generator:
````python
gen = LFRGenerator.fit(G_nk, scale=2)
print(type(gen)) #reveals its class <class 'networkit.generators.LFRGenerator'>
print(dir(gen)) # to list all methods of the class in which are in this particular case
  # ['__class__', '__delattr__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', 
  #'__gt__', '__hash__', '__init__', '__init_subclass__', '__le__', '__lt__', '__ne__', '__new__', 
  #'__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__setstate__', '__sizeof__', '__str__', 
  #'__subclasshook__', 'fit', 'generate', 'generatePowerlawCommunitySizeSequence', 
  #'generatePowerlawDegreeSequence', 'getGraph', 'getPartition', 'hasFinished', 'isParallel', 'params', 
  #'paths', 'run', 'setCommunitySizeSequence', 'setDegreeSequence', 'setMu', 'setMuWithBinomialDistribution',
  # 'setPartition', 'setPathToReferenceImplementationDir', 'toString']
````

That showed us that there is `.run()`, `.generate()` and a `.getGraph()` function.

Much easier is it to look at the test-suite. :https://github.com/networkit/networkit/blob/813f424ede5d6eef67eccbe1f441b0fd7c290198/networkit/test/test_generators.py#L5-L20

wich concludes to this snippet:
```python
gen = LFRGenerator.fit(G_nk, scale=2)
G2 = gen.generate()
print(G2.numberOfNodes())
```

But this results sometimes to `RuntimeError: Graph not realizable, internal degree too large for the assigned community`

and if you want to know the node count before the generation, you cold use the `.toString()` function,
wich is the only way to retrieve the internal value of n wich represents the node count in the C++ class.

`.toString()` returns `"LFRGenerator(40)"` and with a simple regex you could retrieve .

I hope it helps.

**Comment by DeepaMahm** [comment]
Hello @bielefem 

That was really really helpful! Thanks a ton!

`But this results sometimes to RuntimeError: Graph not realizable, internal degree too large for the assigned community`

I'm not sure if I am right. I noticed that this error occurs when the degree of nodes in the input graph ranges from 4 to 6.
I changed the input to a graph with all nodes having a degree < 4 and the scaling works great!

**Comment by DeepaMahm** [comment]
Hi @bielefem 
Each time I run 

Gk = LFRGenerator.fit(G_nk, scale=2).generate()

a different set of edges are returned as output. Is it possible to return the same set of edges always? For instance, while running 
random generators in `networkx` there is a `seed` argument.

Many thanks

**Comment by angriman** [comment]
@DeepaMahm you can use the `nk.setSeed` function before you generate the graph e.g.
```
nk.setSeed(1, True)
Gk = nk.generators.LFRGenerator.fit(G_nk, scale=2).generate()
```
 more on that [here](https://networkit.github.io/dev-docs/python_api/networkit.html?highlight=setseed#networkit.setSeed).

**Comment by DeepaMahm** [comment]
Hi @angriman 
Thanks a lot for the suggestion. I tried,

```
import networkit as nk
import networkx as nx

G = nx.Graph()

G.add_edges_from([ed_ls](https://github.com/DeepaMahm/misc/blob/master/out.txt))
G_nk = nk.nxadapter.nx2nk(G)
# fix seed
nk.setSeed(1, True)
G_nk_scaled = nk.generators.LFRGenerator.fit(G_nk, scale=5).generate()
print(G_nk_scaled.edges())
```

and this returns different edge list for each run.  Could you please have a look?

**Comment by DeepaMahm** [comment]
Hi All,

This is a polite reminder

**Comment by angriman** [comment]
AFAICS `LFRGenerator` is not deterministic when it runs with multiple threads. A way to get a deterministic output is to set the number of threads to 1. You can do this with `nk.engineering.setNumberOfThreads(1)`.

**Comment by michitux** [comment]
@angriman is right, already the fit function is not deterministic with multiple threads.

Concerning the error `Graph not realizable, internal degree too large for the assigned community`, is there any small test instance where this occurs that you can provide together with a seed where this happens? In theory, this should not happen and in particular not with such small degrees. Is it possible that your graph has self-loops? Self-loops are not supported and could explain this error but a check for them is unfortunately missing.

**Comment by DeepaMahm** [comment]
Hi @michitux 
Thanks a lot for the clarification.

> Graph not realizable, internal degree too large for the assigned community
> is there any small test instance where this occurs that you can provide together with a seed where this happens?

Sure, could you please check [this](https://github.com/networkit/networkit/issues/600#issuecomment-702494070) ?

**Comment by michitux** [comment]
Okay, now I understand what is happening. The problem is that the fit function uses `LocalPartitionCoverage` to determine for every node the fraction of its degree that should be inside the community. For this, `LocalPartitionCoverage` uses edge weights. Our implementation of ReCoN, however, does not support edge weights and just rounds these fractions to an internal degree, which can then be larger than the community size.

The error goes away when you make the graph unweighted:

```python
from networkit import graphtools
G_nk_scaled = LFRGenerator.fit(graphtools.toUnweighted(G_nk), scale=2)
print(G_nk_scaled.generate())
```

gives the expected result `NetworKit.Graph(name=G#109, n=40, m=60)`.

The whole network replication code has been written with unweighted, undirected graphs without self-loops in mind. We should have a look at this code again and add checks to detect graphs that cannot be handled by the generator. I think it makes sense to simply exclude them and not to ignore, e.g., the edge weights, as otherwise users might assume edge weights were considered by the generator.

**Comment by DeepaMahm** [comment]
@michitux 
> The problem is that the fit function uses `LocalPartitionCoverage` to determine for every node the fraction of its degree that should be inside the community. For this, `LocalPartitionCoverage` uses edge weights.

Thanks a lot for the details. This is something that I learnt today.



> I think it makes sense to simply exclude them and not to ignore

Thanks a lot for the word of caution. This was the first example that I posted and I wasn't aware that the generator doesn't handle unweighted graphs at that point in time. 

Thanks again for the wonderful support, your time and kind attention.

**Comment by avinash-cn** [comment]
Hi @michitux I am using a graph that is unweighted, undirected, has no self-loops, and still, I get this error `Graph not realizable, internal degree too large for the assigned community`.

Below is the description of my original graph (output of `nk.overview(G)`):

```
Network Properties:
nodes, edges			8415, 91071
directed?			False
weighted?			False
isolated nodes			0
self-loops			0
density				0.002572
clustering coefficient		0.668868
min/max/avg degree		1, 3982, 21.644920
degree assortativity		-0.107006
number of connected components	1
size of largest component	8415 (100.00 %)
```

What could be the reason here? And how can I resolve this?

Thanks.
